src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -1601,24 +1689,31 @@ void c_typecheck_baset::typecheck_expr_trinary(if_exprt &expr)
           operands[1].type().id()!=ID_pointer)
     implicit_typecast(operands[1], operands[2].type());
 
+  auto compile_time_null_pointer = [](const exprt &e, const namespacet &ns) {
+    if(!is_compile_time_constantt(ns)(e))
+      return false;
+    auto s = simplify_expr(e, ns);
```

Comment:
I'd be worried about the use of the simplifier: This may prevent us in the future from making the simplifier smarter, as that might then make additional expressions constant that aren't meant to be.

Reply:
While I agree that we have this problem at the moment (i.e., without the changes proposed in this PR), doesn't the use of `is_compile_time-constantt` before invoking the simplifier address this risk?

Reply:
Would we dare to assert that `s` is a constant?

Reply:
No: `s` could be address-of (which is a permissible constant expression under C11 6.6).

Reply:
Can we assert `is_compile_time_constantt(ns)(s)`?

Reply:
Added!